File: vulns/IDOR/samples/js/parche.js
const express = require('express');
const router = express.Router();
const Invoice = require('../models/Invoice');


router.get('/api/invoices/:invoiceId', authMiddleware, async (req, res) => {
     try {
        // La consulta incluye el ownerId de la sesión como filtro obligatorio.
        // Si la factura no pertenece al usuario, findOne devuelve null
        const invoice = await Invoice.findOne({
            _id: req.params.invoiceId,
            ownerId: req.user.id   // se obtiene del token/sesión verificada
        });

        if (!invoice) {
            return res.status(404).json({ error: 'Invoice not found' });
        }

        return res.json(invoice);
    } catch (err) {
        return res.status(500).json({ error: 'Server error' });
    }
});

File: vulns/IDOR/samples/go/parche.go
package main

import (
    "encoding/json"
    "errors"
    "net/http"
    "strconv"

    "github.com/gorilla/mux"
)

type contextKey string

const userIDKey contextKey = "userID"

func getOrderHandler(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
    orderID, err := strconv.Atoi(vars["orderID"])
    if err != nil {
        http.Error(w, "invalid id", http.StatusBadRequest)
        return
    }

    // La identidad del usuario se obtiene del contexto
    userID, ok := r.Context().Value(userIDKey).(int)
    if !ok {
        http.Error(w, "unauthorized", http.StatusUnauthorized)
        return
    }

	// Autorización aplicada como filtro en la query: si la orden no
    // pertenece al usuario, la consulta no devuelve resultado.
    order, err := db.GetOrderByIDAndUser(orderID, userID)
    if err != nil {
        http.Error(w, "order not found", http.StatusNotFound)
        return
    }

    json.NewEncoder(w).Encode(order)
}

// La consulta para el GetOrder lleva verificacion dentro de sus propios datos